### PR TITLE
Import Senaite and Bika registry records

### PR DESCRIPTION
**Pull request by juangallostra**
**Linked:** https://github.com/senaite/senaite.jsonapi/pull/5, https://github.com/senaite/senaite.api/pull/3
## Current behaviour
Registry records related with Senaite or Bika are not being imported to the target instance
## Expected behaviour after this PR
Registry records related with Senaite and Bika are being imported to the target instance

### CODE REVIEW

File: src/senaite/sync/browser/views.py [hunk 1]
```diff
@@ -427,6 +452,14 @@ def get_authenticated_user(self):
         """
         return self.get_first_item("users/current")
 
+    def get_senaite_registry_records(self):
+        """Return the values of the registry records
+        associated to bika or senaite in the source instance
+
+        :returns:
+        """
+        return self.get_json("registry/senaite-config-registry")
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Here you could do e.g. `self.get_json("registry/bika")` and `self.get_json("registry/senaite")` when you changed the `senaite.jsonapi` accordingly

**Review by ramonski — CHANGES_REQUESTED** [review]

File: src/senaite/sync/browser/views.py [hunk 2]
```diff
@@ -148,11 +151,29 @@ def __call__(self):
             self.fetch_users(domain)
             # Start the fetch process beginning from the portal object
             self.fetch_data(domain, uid="0")
+            # Fetch registry records that contain the word bika or senaite
+            self.fetch_registry_records(domain, ["bika", "senaite"])
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Better name the keyword arguments properly, e.g.: `self.fetch_registry_records(domain, keys=["bika", "senaite"])`

File: src/senaite/sync/browser/views.py [hunk 3]
```diff
@@ -355,6 +375,31 @@ def translate_path(self, path):
         remote_portal_id = path.split("/")[1]
         return path.replace(remote_portal_id, portal_id)
 
+    def fetch_registry_records(self, domain, keywords=None):
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Change `keywords` to`keys`

File: src/senaite/sync/browser/views.py [hunk 4]
```diff
@@ -355,6 +375,31 @@ def translate_path(self, path):
...
+        # If a list of keywords is passed then retrieve only the records associated
+        # to those keywords
+        else:
+            for keyword in keywords:
+                records = self.get_registry_records_by_keyword(keyword)
+                retrieved_records[keyword] = records
+        # Once all the records of interest have been fetched store them like:
+        # storage["registry"][keyword][record] = record_data
+        for keyword in retrieved_records.keys():
+            if retrieved_records[keyword]["items"][0]:
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Reduce complexity of the following code (level of code indendation) by this:

```python
if not retrieved_records[keyword]["items"]:
    continue
registry_store[keyword] = OOBTree()
[...]
```

File: src/senaite/sync/browser/views.py [hunk 5]
```diff
@@ -355,6 +375,31 @@ def translate_path(self, path):
...
+        """Fetch configuration registry records of interest (those associated
+        to the keywords passed) from source instance
+        """
+        logger.info("*** FETCH REGISTRY RECORDS {} ***".format(domain))
+        storage = self.get_storage(domain=domain)
+        registry_store = storage["registry"]
+        retrieved_records = {}
+        # If keyword is None retrieve the whole registry
+        if keywords is None:
+            retrieved_records["all"] = self.get_registry_records_by_keyword()
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Refactor like this to make it more readable (maybe) 🤔

```python
if keys is None:
    keys = []
for key in keys:
    records[key] = self.get_registry_records_by_key(key)
    [...]
else:
    records["all"] = self.get_registry_records_by_key(key)
```

I would also refactor `self.get_registry_records_by_key(key)` to return the contents of `items` immediately

File: src/senaite/sync/browser/views.py [hunk 6]
```diff
@@ -427,6 +472,16 @@ def get_authenticated_user(self):
...
 
+    def get_registry_records_by_keyword(self, keyword=None):
+        """Return the values of the registry records
+        associated to the specified keyword in the source instance.
+        If keyword is None it returns the whole registry
+        """
+        if keyword is None:
+            return self.get_json("registry")
+
+        return self.get_json("registry/{}".format(keyword))
```
Comment (ramonski) [inline src/senaite/sync/browser/views.py]: Refactor this function to return immediately the registry `items`. Maybe you can use the `get_items` method or even the `yield_items` to get the items of all pages.

**Review by ramonski — CHANGES_REQUESTED** [review]

File: src/senaite/sync/browser/views.py (hunk 6 above)
Comment (juangallostra) [reply src/senaite/sync/browser/views.py]: @ramonski The problem we will face by doing this is that then we won't have access to the api url for those items

Comment (juangallostra) [reply src/senaite/sync/browser/views.py]: @ramonski I have refactored the method to directly return `items` and for now I have removed the API url column for registry records. If you know of any way I can get the api url from within `sync` I will use it to get them

File: src/senaite/sync/browser/views.py (hunk 5 above)
Comment (juangallostra) [reply src/senaite/sync/browser/views.py]: @ramonski I tried to refactor it as specified but I was facing a problem where the `else` clause was being always executed independently of the value of `keys`. As a results I was getting not only the registry records associated to bika or senaite but all of them.

**Review by ramonski — APPROVED** [review]

**Pull request merged**